File: PaymentGatewayUI - v1.0/app/src/main/java/com/example/paymentgatewayui/FailureActivity.java
package com.example.paymentgatewayui;

import android.content.Intent;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.view.View;
import android.widget.Button;

import androidx.appcompat.app.AppCompatActivity;

public class FailureActivity extends AppCompatActivity {

    private Handler handler = new Handler(Looper.getMainLooper());
    private Runnable redirectRunnable;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.failure);

        Button retryButton = findViewById(R.id.tryAgainButton);

        retryButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                cancelRedirect();
                goToMain();
            }
        });

        redirectRunnable = this::goToMain;
        handler.postDelayed(redirectRunnable, 3000); // 3 seconds
    }

    private void goToMain() {
        Intent intent = new Intent(FailureActivity.this, MainActivity.class);
        intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP | Intent.FLAG_ACTIVITY_NEW_TASK);
        startActivity(intent);
        finish();
    }

    private void cancelRedirect() {
        if (redirectRunnable != null) {
            handler.removeCallbacks(redirectRunnable);
        }
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        cancelRedirect(); // Avoid memory leak
    }
}

File: PaymentGatewayUI - v2.0/app/src/main/java/com/example/paymentgatewayui/FailureActivity.kt
package com.example.paymentgatewayui

import android.content.Intent
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.widget.Button
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity

class FailureActivity : AppCompatActivity() {
    private val handler = Handler(Looper.getMainLooper())
    private var redirectRunnable: Runnable? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_failure)

        // Get error message passed from PaymentMethodActivity
        val errorMessage = intent.getStringExtra("errorMessage") ?: "An unexpected error occurred"

        // Set error message to TextView
        val errorTextView: TextView = findViewById(R.id.failureMessageText)
        errorTextView.text = errorMessage

        val retryButton: Button = findViewById(R.id.tryAgainButton)
        retryButton.setOnClickListener {
            cancelRedirect()
            navigateBackToTransactionType()
        }

        // Auto-redirect after 5 seconds
        redirectRunnable = Runnable {
            navigateBackToTransactionType()
        }
        handler.postDelayed(redirectRunnable!!, 5000)
    }

    private fun navigateBackToTransactionType() {
        val intent = Intent(this, TransactionTypeActivity::class.java)
        intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP or Intent.FLAG_ACTIVITY_NEW_TASK)
        startActivity(intent)
        finish()
    }

    private fun cancelRedirect() {
        redirectRunnable?.let {
            handler.removeCallbacks(it)
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        cancelRedirect()
    }
}